### PR TITLE
fix: stabilize team runtime flows

### PR DESCRIPTION
**Pull request by hawkingrei**
## Summary

This PR is the dedicated fix stream for Team runtime issues encountered during active use.

The first fix in this branch addresses Team worker agent creation from `Teams -> Add Agent`:

- default worker forge config now follows Team runtime policy and uses `create_worktree`
- worker forge defaults infer `worktree_repo` from existing Team member runtime hints when available
- the modal no longer opens in the invalid `use_existing + empty workdir` state that blocked submission before any create-agent request was sent

Future Team runtime bug fixes will continue landing on this branch/PR until the current stabilization pass is complete.

## Validation

- `cd web && npx vitest run src/pages/team/forge_helpers.test.ts`
- `cd web && npm run lint -- src/pages/team/forge_helpers.ts src/pages/team/forge_helpers.test.ts src/pages/team_page.tsx`
- `cd web && npm run build`

## Notes

- Chrome DevTools MCP baseline was captured on `https://agenthub.hawkingrei.com/teams/...` before the fix:
  - worker `Add Agent` opened with an empty `Workdir`
  - clicking `Create Agent` did not issue a backend create request
  - the page surfaced `Agent workdir is required`
- Deployed-domain regression still needs verification after rollout; tracked in `docs/todo.md`.

### CODE REVIEW

File: src/team/manager.rs [hunk 1]
```diff
@@ -2125,6 +2129,126 @@ impl TeamManager {
...
+        if let Some(existing) = self
+            .get_latest_shared_thread_mailbox_run(team_id, task_id)
+            .await?
+        {
+            return Ok(existing);
+        }
+
+        let run_id = Uuid::new_v4().to_string();
+        let context_id = format!("shared-thread-mailbox:{task_id}");
+        let now = Utc::now().timestamp();
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply src/team/manager.rs:2209]: Done. `ensure_shared_thread_mailbox_run()` now uses a deterministic hidden run id (`shared-thread-mailbox:{team_id}:{task_id}`) plus `INSERT OR IGNORE`, and only appends bootstrap events on the first insert. I also added an idempotency regression test so concurrent callers converge on one hidden mailbox run instead of accumulating duplicates.

File: src/api/teams.rs [hunk 2]
```diff
@@ -2360,6 +2362,42 @@ async fn load_latest_active_run_for_team(
...
+    if task.title.trim().eq_ignore_ascii_case("all") {
+        return true;
+    }
+    task.context
+        .as_object()
+        .and_then(|obj| obj.get("bootstrap_kind"))
+        .and_then(Value::as_str)
+        .map(str::trim)
+        .is_some_and(|value| value.eq_ignore_ascii_case(TEAM_SHARED_THREAD_BOOTSTRAP_KIND))
+}
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply src/api/teams.rs:2464]: Agreed on the maintainability concern. I kept cross-language sharing out of scope for this PR, but I did centralize the frontend shared-thread identifiers into `DEFAULT_TEAM_THREAD_TITLE` / `DEFAULT_TEAM_THREAD_BOOTSTRAP_KIND` and switched the Team page/tests to consume those constants instead of repeating the raw strings at multiple call sites.

File: web/src/pages/team_page.tsx [hunk 3]
```diff
@@ -1894,21 +1910,42 @@ export function TeamPage(props: TeamPageProps) {
...
+          }),
+          taskRecord && isSharedThreadTask(taskRecord)
+            ? api.getTeamTask(props.token, teamId, taskId)
+            : Promise.resolve(null),
+        ]);
         setTaskMessages(messages);
+        const conversationRunId = taskDetail?.latest_run?.id?.trim() ?? "";
+        if (conversationRunId) {
+          const conversationSnapshot = await api.getTeamRunSnapshot(props.token, conversationRunId, {
+            event_limit: 1,
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply web/src/pages/team_page.tsx]: Fixed later in this branch. `refreshTaskMessages` now lives in `useTeamConversationActions` behind a monotonic request-sequence plus `{teamId, taskId}` scope guard, and it only applies `setTaskMessages` / `setConversationMailboxMessages` if the response still matches the current Team and conversation selection.

File: src/state.rs [hunk 4]
```diff
@@ -174,6 +195,37 @@ impl AppState {
...
+
+    fn ensure_global_gitignore_agenthubmemory() -> anyhow::Result<()> {
+        let Some(home) = std::env::var_os("HOME").filter(|home| !home.is_empty()) else {
+            return Ok(());
+        };
+
+        let home_path = PathBuf::from(home);
+        std::fs::create_dir_all(&home_path)?;
+        let gitignore_path = home_path.join(GLOBAL_GITIGNORE_FILENAME);
+        let existing = match std::fs::read_to_string(&gitignore_path) {
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply src/state.rs]: Fixed later in this branch. Startup now appends `.agenthubmemory` to both `$HOME/.gitignore_global` and the Git default user ignore path (`$XDG_CONFIG_HOME/git/ignore` or `~/.config/git/ignore`), so the protection no longer depends on `core.excludesfile` pointing at `~/.gitignore_global`.

File: web/src/pages/team_member_acp_panel.tsx [hunk 5]
```diff
@@ -64,30 +68,62 @@ export function TeamMemberAcpPanel(props: TeamMemberAcpPanelProps) {
...
+  const acpConversation = useAcpConversation({
+    acpView,
+    activeAgent: selectedMemberId.trim() || null,
+    activeSessionId: selectedSessionId ?? null,
+    acpTab: "conversation",
+    eventMeta: conversationEventMeta,
+    isAgentActive: Boolean(selectedSessionId),
+    onLoadOlder: () => {
+      void onLoadOlder();
+    },
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply web/src/pages/team_member_acp_panel.tsx:103]: Fixed later in this branch. `TeamMemberAcpPanel` now passes `effectiveAcpTab` through to `useAcpConversation`, so the conversation auto-load, scroll, and jump behavior only stays active while the Conversation tab is actually selected.

File: src/team/manager.rs [hunk 6]
```diff
@@ -1171,7 +1174,8 @@ impl TeamManager {
         for row in rows {
             runs.push(parse_team_run_row(&row)?);
         }
-        self.hydrate_run_summaries(runs).await
+        let runs = self.hydrate_run_summaries(runs).await?;
+        Ok(filter_visible_team_runs(runs))
     }
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply src/team/manager.rs:1212]: Fixed later in this branch. `list_runs` now excludes hidden `shared_thread_mailbox` runs in the SQL `WHERE` clause before `LIMIT` and cursor pagination, and `list_runs_supports_status_filter_and_cursor` keeps the visible pagination contract covered.

File: web/src/pages/team/forge_helpers.ts [hunk 7]
```diff
@@ -36,19 +36,63 @@ export type TeamForgeDefaults = {
...
+    return null;
+  }
+  return value as Record<string, unknown>;
+}
+
+function readRuntimePath(value: unknown, key: "worktree_repo" | "workdir"): string {
+  const record = asObjectRecord(value);
+  const raw = record?.[key];
+  return typeof raw === "string" ? normalizeWorkdirInput(raw) : "";
+}
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply web/src/pages/team/forge_helpers.ts:64]: Fixed later in this branch. resolveTeamForgeDefaults() now fills worktreeRepo via resolveTeamForgeWorktreeRepo(teamSpec), while the worker Workdir resolver only consumes runtime workdir hints. That keeps the modal fields split by intent instead of leaving worktreeRepo permanently empty.

File: src/api/teams.rs [hunk 8]
```diff
@@ -2360,6 +2362,42 @@ async fn load_latest_active_run_for_team(
...
+    }
+    if !is_shared_thread_task(task) {
+        return Ok(None);
+    }
+    let conversation = state
+        .teams
+        .get_task_conversation(&task.id)
+        .await
+        .map_err(map_team_internal_error)?;
+    let run = state
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply src/api/teams.rs]: Fixed later in this branch. The shared-thread/no-active-run path now threads message.conversation_id into resolve_task_message_mailbox_run(...), so the mailbox-run resolver no longer re-queries team_conversations on every send.

File: web/src/pages/team/use_team_conversation_effects.ts [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+  useEffect(() => {
+    const conversationId = selectedConversationId?.trim() ?? "";
+    if (!selectedTeamId || !conversationId) {
+      setTaskMessages([]);
+      setConversationMailboxMessages([]);
+      return;
+    }
+    void refreshSelectedConversation();
+  }, [
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply web/src/pages/team/use_team_conversation_effects.ts:92]: Fixed later in this branch. useTeamConversationEffects now queues a follow-up refresh when selection changes mid-flight, and use_team_conversation_effects.test.tsx keeps that regression covered with the selection-change queue test.

File: web/src/pages/team/page_helpers.ts [hunk 10]
```diff
@@ -225,9 +262,20 @@ export function upsertEventList(
...
+  const scopedPrev =
+    sessionId == null
+      ? prev
+      : prev.filter((event) => (event.session_id ?? null) === sessionId);
+  if (mode === "replace") {
+    if (next.length === 0) {
+      return [];
+    }
+    return mergeOutputsPreserveHistory(scopedPrev, next, true);
+  }
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply web/src/pages/team/page_helpers.ts:314]: Fixed. upsertAgentEventList(..., replace) now preserves history only when a concrete sessionId is present; without a session id it restores the original replace semantics and drops older events. I added a regression test for the no-session case in page_helpers.test.ts.

File: web/src/pages/team/forge_helpers.ts [hunk 11]
```diff
@@ -146,13 +232,13 @@ export function resolveTeamForgeDefaults({
...
-            "",
+            resolveTeamForgeWorkerWorkdir(teamSpec),
             "use_existing",
             defaultWorktreeRoot,
             DEFAULT_WORKTREE_ROOT
           ),
     worktreeMode: "use_existing",
-    worktreeRepo: "",
+    worktreeRepo: resolveTeamForgeWorktreeRepo(teamSpec),
     worktreeRef: "",
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply web/src/pages/team/forge_helpers.ts:245]: This was intentionally superseded later in the branch. After validating the worker flow on real Team setups, the default moved back to use_existing so teams can point workers at an existing repo or workdir first; create_worktree remains available as an explicit advanced choice. The code, tests, and docs now reflect that later product decision, so I am not switching the default back in this PR.

File: web/src/pages/team/page_helpers.ts [hunk 12]
```diff
@@ -225,16 +263,73 @@ export function upsertEventList(
...
   }
   return [...byId.values()].sort((a, b) => a.event_id - b.event_id);
 }
 
+function sameConversationPayload(left: unknown, right: unknown): boolean {
+  if (left === right) {
+    return true;
+  }
+  return JSON.stringify(left) === JSON.stringify(right);
+}
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply web/src/pages/team/page_helpers.ts]: Fixed. Team conversation merges no longer stringify arbitrary payloads on every refresh. The merge path now reuses cached messages by stable scalar identity, which matches the backend append-only message model and removes the extra CPU and allocation cost.

File: web/src/pages/team/forge_helpers.ts [hunk 13]
```diff
@@ -146,13 +235,13 @@ export function resolveTeamForgeDefaults({
...
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply web/src/pages/team/forge_helpers.ts:245]: This was intentionally superseded later in the branch. After validating the worker flow on real Team setups, the default moved back to use_existing so teams can point workers at an existing repo or workdir first; create_worktree remains available as an explicit advanced choice. The code, tests, and docs now reflect that later product decision, so I am not switching the default back in this PR.

File: src/sse.rs [hunk 14]
```diff
@@ -98,9 +104,46 @@ async fn sse_agents(
...
+    Path((team_id, task_id)): Path<(String, String)>,
+    Query(query): Query<SseTokenQuery>,
+) -> impl IntoResponse {
+    let user = match state.auth.validate_session(&query.token).await {
+        Ok(user) => user,
+        Err(_) => return (StatusCode::UNAUTHORIZED, "unauthorized").into_response(),
+    };
+    if load_team_for_user(&state, &team_id, &user).await.is_err() {
+        return (StatusCode::NOT_FOUND, "team not found").into_response();
+    }
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply src/sse.rs]: Fixed. Team task-message SSE now preserves real internal failures as 500 responses instead of collapsing everything into 404 team not found. I also added team_task_messages_sse_preserves_internal_errors_as_500 to keep that behavior covered.

File: src/state.rs [hunk 15]
```diff
@@ -174,14 +205,114 @@ impl AppState {
...
     }
+
+    fn ensure_global_gitignore_agenthubmemory() -> anyhow::Result<()> {
+        let Some(home) = std::env::var_os("HOME").filter(|home| !home.is_empty()) else {
+            return Ok(());
+        };
+
+        let home_path = PathBuf::from(home);
+        std::fs::create_dir_all(&home_path)?;
+        for gitignore_path in resolve_global_gitignore_paths(&home_path) {
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply src/state.rs:215]: Fixed. ensure_global_gitignore_agenthubmemory() no longer calls create_dir_all(HOME); it only appends to the concrete ignore-file paths, and append_gitignore_entry() remains responsible for creating the parent directories it actually needs.

File: web/src/pages/team/forge_helpers.ts [hunk 16]
```diff
@@ -146,13 +235,13 @@ export function resolveTeamForgeDefaults({
...
         : resolveWorkdirForModalOpen(
-            "",
+            resolveTeamForgeWorkerWorkdir(teamSpec),
             "use_existing",
             defaultWorktreeRoot,
             DEFAULT_WORKTREE_ROOT
           ),
     worktreeMode: "use_existing",
-    worktreeRepo: "",
+    worktreeRepo: resolveTeamForgeWorktreeRepo(teamSpec),
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply web/src/pages/team/forge_helpers.ts:244]: This was intentionally superseded later in the branch. After validating the worker flow on real Team setups, the default moved back to use_existing so teams can point workers at an existing repo or workdir first; create_worktree remains available as an explicit advanced choice. The code, tests, and docs now reflect that later product decision, so I am not switching the default back in this PR.

File: web/src/pages/team/page_helpers.ts [hunk 17]
```diff
@@ -225,16 +299,73 @@ export function upsertEventList(
...
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply web/src/pages/team/page_helpers.ts]: Fixed. Team conversation merges no longer stringify arbitrary payloads on every refresh. The merge path now reuses cached messages by stable scalar identity, which matches the backend append-only message model and removes the extra CPU and allocation cost.

File: src/state.rs [hunk 18]
```diff
@@ -24,9 +28,18 @@ pub struct AppState {
...
+const DEFAULT_GIT_IGNORE_SUBPATH: &str = "git/ignore";
 
 impl AppState {
     pub async fn init(config: agenthub_config::AppConfig) -> anyhow::Result<Self> {
+        if let Err(error) = Self::ensure_global_gitignore_agenthubmemory() {
+            tracing::warn!(
+                ?error,
+                "failed to ensure global gitignore entry for .agenthubmemory"
+            );
+        }
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply src/state.rs:42]: I kept the startup write itself in place because the product requirement here is to auto-protect .agenthubmemory for local project workspaces. I did trim the side effects in this follow-up by removing the blanket HOME directory creation and limiting writes to the concrete Git ignore files. A broader deployment-mode or config gate is a separate follow-up if we later need to distinguish local interactive runs from hosted service deployments.

File: src/team/runtime.rs [hunk 19]
```diff
@@ -42,12 +42,19 @@ struct TeamRuntimeMemberSpec {
...
     name: Option<String>,
     workdir: Option<String>,
     worktree_repo: Option<String>,
     worktree_ref: Option<String>,
+    #[serde(default)]
+    agent_loop_enabled: Option<bool>,
+    #[serde(default)]
+    agent_loop_idle_seconds: Option<i64>,
+    #[serde(default)]
+    agent_loop_prompt: Option<String>,
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply src/team/runtime.rs]: Fixed. I removed the struct-level allow(dead_code) and narrowed the allowance to the currently-unused fields only, so we still get dead-code signal if the hint struct itself or the active runtime fields stop being used.

File: web/src/pages/team/forge_helpers.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (hawkingrei) [reply web/src/pages/team/forge_helpers.ts:245]: This was intentionally superseded later in the branch. After validating the worker flow on real Team setups, the default moved back to use_existing so teams can point workers at an existing repo or workdir first; create_worktree remains available as an explicit advanced choice. The code, tests, and docs now reflect that later product decision, so I am not switching the default back in this PR.

File: web/src/pages/team/page_helpers.ts [hunk 20]
```diff
@@ -225,16 +299,73 @@ export function upsertEventList(
...
+  return (
+    left.message_id === right.message_id &&
+    left.conversation_id === right.conversation_id &&
+    left.task_id === right.task_id &&
+    left.from_actor_id === right.from_actor_id &&
+    (left.to_actor_id ?? null) === (right.to_actor_id ?? null) &&
+    left.route === right.route &&
+    left.created_at === right.created_at &&
+    sameConversationPayload(left.payload, right.payload)
+  );
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply web/src/pages/team/page_helpers.ts]: Fixed. Team conversation merges no longer stringify arbitrary payloads on every refresh. The merge path now reuses cached messages by stable scalar identity, which matches the backend append-only message model and removes the extra CPU and allocation cost.

File: src/team/runtime.rs [hunk 21]
```diff
@@ -42,12 +42,19 @@ struct TeamRuntimeMemberSpec {
...
 }
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply src/team/runtime.rs]: Fixed. I removed the struct-level allow(dead_code) and narrowed the allowance to the currently-unused fields only, so we still get dead-code signal if the hint struct itself or the active runtime fields stop being used.

File: web/src/pages/team/forge_helpers.ts [hunk 22]
```diff
@@ -146,13 +235,13 @@ export function resolveTeamForgeDefaults({
...
     worktreeRef: "",
   };
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply web/src/pages/team/forge_helpers.ts:246]: This was intentionally superseded later in the branch. After validating the worker flow on real Team setups, the default moved back to use_existing so teams can point workers at an existing repo or workdir first; create_worktree remains available as an explicit advanced choice. The code, tests, and docs now reflect that later product decision, so I am not switching the default back in this PR.

File: web/src/pages/team/create_helpers.ts [hunk 23]
```diff
@@ -217,6 +220,148 @@ export function appendTeamMemberToSpec(
...
+    model: draft.model.trim() || undefined,
+    prompt,
+    skills: normalizedSkills,
+    runtime: {
+      ...asObjectRecord(existing.runtime),
+      agent_loop_enabled: draft.agent_loop_enabled || undefined,
+      agent_loop_idle_seconds:
+        draft.agent_loop_idle_seconds.trim() !== "" && Number.isFinite(parsedLoopIdleSeconds)
+          ? parsedLoopIdleSeconds
+          : undefined,
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply web/src/pages/team/create_helpers.ts]: Fixed. updateTeamMemberProfileInSpec() now accepts agent_loop_idle_seconds only when the trimmed value is strictly numeric; partially numeric input like 900abc is treated as unset. I added a regression test for that case in create_helpers.test.ts.

File: web/src/pages/team/create_helpers.ts [hunk 24]
```diff
@@ -217,6 +220,152 @@ export function appendTeamMemberToSpec(
...
+  return readRuntimeRecord(member)?.agent_loop_enabled === true;
+}
+
+function readRuntimeLoopIdleSeconds(member: Record<string, unknown>): string {
+  const value = readRuntimeRecord(member)?.agent_loop_idle_seconds;
+  if (typeof value === "number" && Number.isFinite(value) && value > 0) {
+    return String(Math.trunc(value));
+  }
+  return "";
+}
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply web/src/pages/team/create_helpers.ts]: Fixed. Draft hydration now drops out-of-range loop idle seconds and only surfaces values in the same 10..=86400 window the backend accepts, so stale spec or runtime hints no longer prefill non-actionable loop timeouts.

File: web/src/pages/team/create_helpers.ts [hunk 25]
```diff
@@ -217,6 +220,152 @@ export function appendTeamMemberToSpec(
...
+    skills: normalizedSkills,
+    runtime: {
+      ...asObjectRecord(existing.runtime),
+      agent_loop_enabled: draft.agent_loop_enabled || undefined,
+      agent_loop_idle_seconds:
+        loopIdleRaw !== "" && Number.isFinite(parsedLoopIdleSeconds)
+          ? parsedLoopIdleSeconds
+          : undefined,
+      agent_loop_prompt: draft.agent_loop_prompt.trim() || undefined,
+    },
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply web/src/pages/team/create_helpers.ts:378]: Fixed. updateTeamMemberProfileInSpec() now validates agent_loop_idle_seconds against the same 10..=86400 bounds as the backend; out-of-range values are treated as unset instead of being written into runtime.agent_loop_idle_seconds.
